### ISSUE TITLE
ParralelDecompiler Faster?

### ISSUE THREAD

**Issue by clearbluejar** [clearbluejar/ghidrecomp#1]
Instead of using python threading, would it be faster to use ParallelDecompiler?

```python
        from decompiler import MyDecompileConfigurer
        from ghidra.app.decompiler.parallel import ParallelDecompiler
        from ghidra.app.decompiler.parallel import DecompilerCallback
        from ghidra.app.decompiler.parallel import DecompileConfigurer

        configurer = MyDecompileConfigurer()
        callback = DecompilerCallback(program, DecompileConfigurer)

        ParallelDecompiler.decompileFunctions(callback, program, all_funcs, None, monitor)
```

```python
@JImplements(DecompileConfigurer, deferred=False)
class MyDecompileConfigurer:

    @JOverride
    def configure(self, decompiler: 'ghidra.app.decompiler.DecompInterface'):
        decompiler.toggleCCode(False)
        decompiler.toggleSyntaxTree(True)
        decompiler.setSimplificationStyle("decompile")
        opts = DecompileOptions()
        opts.grabFromProgram(p)
        decompiler.setOptions(opts)
```

**Comment by astrelsky** [comment]
> Instead of using python threading, would it be faster to use ParallelDecompiler?
> 

Yes. "CPython implementation detail: In CPython, due to the [Global Interpreter Lock](https://docs.python.org/3/glossary.html#term-global-interpreter-lock), only one thread can execute Python code at once (even though certain performance-oriented libraries might overcome this limitation)." [source](https://docs.python.org/3/library/threading.html).

You should use the `ParallelDecompiler`. If you need to I've used some Java threads from python using pyhidra before with some success. Synchronization will occur via the GIL when everytime you come from java into python code though.

**Comment by clearbluejar** [comment]
Thanks for this. In my case, I [spin up multiple decompilers](https://github.com/clearbluejar/ghidrecomp/blob/main/ghidrecomp/decompile.py#L25C5-L40C6) and have "multithreaded" python [submit each function to several running decompilers](https://github.com/clearbluejar/ghidrecomp/blob/main/ghidrecomp/decompile.py#L221C1-L229), so perhaps the GIL issue is mitigated?  

I had trouble using pyhidra to instantiate a callback function.  

```python
callback = DecompilerCallback(program, DecompileConfigurer)
````

Which caused `java.lang.InstantiationException: ghidra.app.decompiler.parallel.DecompilerCallback`

**Comment by astrelsky** [comment]
> DecompilerCallback

`DecompilerCallback` is an abstract class which you can't instantiate directly. You must use a subclass of it. Unfortunately there don't appear to be any subclasses with public visibility so you'd need to subclass it yourself and you'd have to do it in Java.

**Comment by astrelsky** [comment]
> Thanks for this. In my case, I [spin up multiple decompilers](https://github.com/clearbluejar/ghidrecomp/blob/main/ghidrecomp/decompile.py#L25C5-L40C6) and have "multithreaded" python [submit each function to several running decompilers](https://github.com/clearbluejar/ghidrecomp/blob/main/ghidrecomp/decompile.py#L221C1-L229), so perhaps the GIL issue is mitigated?
> 

The GIL "issue" is intentional. What I *think* is happening is that when a Java function is called, Jpype releases the GIL which then allows another Python thread to acquire the GIL and run Python code. You will have multiple Python threads, *I think*, you just can't run Python code in multiple threads at the same time in the same Python process/interpreter/whatever.

**Comment by clearbluejar** [comment]
> > DecompilerCallback
> 
> `DecompilerCallback` is an abstract class which you can't instantiate directly. You must use a subclass of it. Unfortunately there don't appear to be any subclasses with public visibility so you'd need to subclass it yourself and you'd have to do it in Java.

I have been trying to keep it strictly python, but in this case would I need to implement the subclass as a plugin?  Pyhidra does has [some support](https://github.com/dod-cyber-crime-center/pyhidra#registering-an-entry-point) to include a plugin, but I have yet to wrap my head around it.

**Comment by astrelsky** [comment]
> > > DecompilerCallback
> > 
> > 
> > `DecompilerCallback` is an abstract class which you can't instantiate directly. You must use a subclass of it. Unfortunately there don't appear to be any subclasses with public visibility so you'd need to subclass it yourself and you'd have to do it in Java.
> 
> I have been trying to keep it strictly python, but in this case would I need to implement the subclass as a plugin? Pyhidra does has [some support](https://github.com/dod-cyber-crime-center/pyhidra#registering-an-entry-point) to include a plugin, but I have yet to wrap my head around it.
> 
> edit: Understanding that it would need to be subclassed in Java, what is the best way to implement it? A plugin, or some 1-off jpype command like `jpype.addClassPath('/my/path/myJar.jar')`?

I don't know which would be best here. I'd say to just do whatever you feel is easiest.
